test/rubygems/test_require.rb

```diff
@@ -677,6 +677,39 @@ def test_require_bundler_with_bundler_version
         end
       end
     end
+
+    def test_no_crash_when_overriding_warn_with_warning_module
+      skip "https://github.com/oracle/truffleruby/issues/2109" if RUBY_ENGINE == "truffleruby"
+
+      Dir.mktmpdir("warn_test") do |dir|
+        File.write(dir + "/main.rb", <<-'RUBY')
+          module Warning
+            def warn(str)
+              super
+            end
+          end
+          warn 'Foo Bar'
+          class C < BasicObject
+            include ::Kernel
+            def some_method
+              warn "warn comes from #{method(:warn).source_location.inspect}"
+            end
+          end
+          C.new.some_method
+        RUBY
+        _, err = capture_subprocess_io do
+          system(*ruby_with_rubygems_in_load_path, "-w", "--disable=gems", "-C", dir, "main.rb")
+        end
+        assert_match(/^Foo Bar\n/, err)
+        assert_match(/^warn comes from/, err)
+        refute_match(%r{^warn comes from .+rubygems/core_ext/kernel_warn.rb}, err)
```

Comment:
The reason for this `refute` vs asserting it comes from core is that's it's not easy to assert it comes from core, because methods defined in Ruby will have a source_location. In Ruby 2.7:
```
$ ruby --disable-gems -e 'p method(:warn).source_location'
["<internal:warning>", 42]
```